runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h

```diff
@@ -25,6 +25,8 @@ namespace atn {
 
     static Ref<const SingletonPredictionContext> create(Ref<const PredictionContext> parent, size_t returnState);
 
+    PredictionContextType getContextType() const override;
```

Comment:
For all classes that are not extended anymore, it would probably be good to use `final` instead of `override`.
Not only is this a good documentation for the reader of the code, but the compiler can also use that for devirtualization if it otherwise can know which class it is dealing with.

Reply:
Unfortunately SingletonPredictionContext is inherited by EmptyPredictionContext, so it cannot be final. However, whether or not we need PredictionContextType::EMPTY I will have to look.

Reply:
I was able to remove EmptyPredictionContext and make SingletonPredictionContext final. However it won't matter much, as most invocations use the base class rather than the derived. So the compiler cannot do much.

Reply:
I usually find it most useful as documentation for the reader of the code, actually.